### PR TITLE
purego: add float32 and float64 support to callbacks

### PR DESCRIPTION
**Pull request by TotallyGamerJet**
This commit adds support for float32 and float64 arguments in C callbacks. It also adds a test to check the maximum number of arguments a callback can have and that it is callback properly by purego. There is also some formatting done to the assembly.

### CODE REVIEW

File: callback_test.go [hunk 1]
```diff
@@ -72,3 +72,30 @@ func buildSharedLib(libFile string, sources ...string) error {
...
+func TestNewCallback(t *testing.T) {
+	// This tests the maximum number of arguments a function to NewCallback can take
+	const (
+		expectCbTotal    = -3
+		expectedCbTotalF = float64(36)
+	)
+	var cbTotal int
+	var cbTotalF float64
+	imp := purego.NewCallback(func(a1, a2, a3, a4, a5, a6, a7, a8, a9 int,
+		f1, f2, f3, f4, f5, f6, f7, f8 float64) {
```
Comment (hajimehoshi) [inline callback_test.go:85]: Can we also test float32?

File: syscall_sysv.go [hunk 2]
```diff
@@ -136,9 +133,34 @@ func callbackWrap(a *callbackArgs) {
 	fnType := fn.Type()
 	args := make([]reflect.Value, fnType.NumIn())
 	frame := (*[callbackMaxFrame]uintptr)(a.args)
+	var floatsN int
+	// offset the integer position by the number
+	// of floatsN because in the frame it starts with the float
+	// registers followed by the integer and then the stack after that.
```
Comment (hajimehoshi) [inline syscall_sysv.go:139]: Is this the explanation about the variable intsN?

Comment (TotallyGamerJet) [reply syscall_sysv.go:136]: yes

File: callback_test.go (hunk 1 above)
Comment (TotallyGamerJet) [reply callback_test.go:85]: done

File: callback_test.go [hunk 3]
```diff
@@ -72,3 +72,57 @@ func buildSharedLib(libFile string, sources ...string) error {
...
+		cbTotalF = f1 + f2 + f3 + f4 + f5 + f6 + f7 + f8
+	})
+	var fn func(a1, a2, a3, a4, a5, a6, a7, a8, a9 int,
+		f1, f2, f3, f4, f5, f6, f7, f8 float64)
+	purego.RegisterFunc(&fn, imp)
+	fn(1, 2, -3, 4, -5, 6, -7, 8, -9,
+		1, 2, 3, 4, 5, 6, 7, 8)
+
+	if cbTotal != expectCbTotal {
+		t.Fatalf("cbTotal not correct got %d but wanted %d", cbTotal, expectCbTotal)
```
Comment (hajimehoshi) [inline callback_test.go:96]: Error?

File: callback_test.go [hunk 4]
```diff
@@ -72,3 +72,57 @@ func buildSharedLib(libFile string, sources ...string) error {
...
+	}
+	if cbTotalF != expectedCbTotalF {
+		t.Fatalf("cbTotalF not correct got %f but wanted %f", cbTotalF, expectedCbTotalF)
```
Comment (hajimehoshi) [inline callback_test.go:99]: ditto

File: callback_test.go [hunk 5]
```diff
@@ -72,3 +72,57 @@ func buildSharedLib(libFile string, sources ...string) error {
...
+		cbTotalF = f1 + f2 + f3 + f4 + f5 + f6 + f7 + f8 + f9
+	})
+	var fn func(a1, a2, a3, a4, a5, a6, a7, a8 int,
+		f1, f2, f3, f4, f5, f6, f7, f8, f9 float32)
+	purego.RegisterFunc(&fn, imp)
+	fn(1, 2, -3, 4, -5, 6, -7, 8,
+		1, 2, 3, 4, 5, 6, 7, 8, 9)
+
+	if cbTotal != expectCbTotal {
+		t.Fatalf("cbTotal not correct got %d but wanted %d", cbTotal, expectCbTotal)
```
Comment (hajimehoshi) [inline callback_test.go:123]: ditto

File: callback_test.go [hunk 6]
```diff
@@ -72,3 +72,57 @@ func buildSharedLib(libFile string, sources ...string) error {
...
+	}
+	if cbTotalF != expectedCbTotalF {
+		t.Fatalf("cbTotalF not correct got %f but wanted %f", cbTotalF, expectedCbTotalF)
```
Comment (hajimehoshi) [inline callback_test.go:126]: ditto

File: callback_test.go [hunk 7]
```diff
@@ -72,3 +72,57 @@ func buildSharedLib(libFile string, sources ...string) error {
 
 	return nil
 }
+
+func TestNewCallback(t *testing.T) {
```
Comment (hajimehoshi) [inline callback_test.go:76]: TestNewCallbackFloat64?

File: callback_test.go [hunk 8]
```diff
@@ -72,3 +72,57 @@ func buildSharedLib(libFile string, sources ...string) error {
...
+
+	if cbTotal != expectCbTotal {
+		t.Fatalf("cbTotal not correct got %d but wanted %d", cbTotal, expectCbTotal)
+	}
+	if cbTotalF != expectedCbTotalF {
+		t.Fatalf("cbTotalF not correct got %f but wanted %f", cbTotalF, expectedCbTotalF)
+	}
+}
+
+func TestNewCallback32(t *testing.T) {
```
Comment (hajimehoshi) [inline callback_test.go:103]: TestNewCallbackFloat32?

**Review by hajimehoshi — COMMENTED** [review]
Thanks!

By the way, Does a calback work even if this mixes float32 and float64 as this arguments?

File: syscall_sysv.go [hunk 9]
```diff
@@ -136,9 +130,33 @@ func callbackWrap(a *callbackArgs) {
 	fnType := fn.Type()
 	args := make([]reflect.Value, fnType.NumIn())
 	frame := (*[callbackMaxFrame]uintptr)(a.args)
+	var floatsN int
+	// intsN starts at numOfFloats because floats occur before integers
+	// in the callbackArgs.args field.
+	var intsN int = numOfFloats
+	// the stack is located in the frame after the floats and integers
+	var stack = numOfIntegerRegisters() + numOfFloats
```
Comment (hajimehoshi) [inline syscall_sysv.go:138]: The comment should explain what the variable `stack` is.

**Comment by hajimehoshi** [comment]
I'm gonna sleep soon so I'll take a look tomorrow. Good night...

**Comment by TotallyGamerJet** [comment]
> By the way, Does a callback work even if this mixes float32 and float64 as this arguments?

Yes! I added a test for that case (TestNewCallbackFloat32AndFloat64)

**Review by hajimehoshi — APPROVED** [review]
LGTM, thanks!
